File: npmjs/pray-cli/src/resolve/dependency-graph.ts
export function findDependencyCycle(
  edges: Map<string, string[]>,
): string[] | undefined {
  const visiting = new Set<string>();
  const visited = new Set<string>();
  const stack: string[] = [];
  for (const name of [...edges.keys()].sort()) {
    if (visited.has(name)) {
      continue;
    }
    const cycle = depthFirstSearch(name, edges, visiting, visited, stack);
    if (cycle) {
      return cycle;
    }
  }
  return undefined;
}

function depthFirstSearch(
  name: string,
  edges: Map<string, string[]>,
  visiting: Set<string>,
  visited: Set<string>,
  stack: string[],
): string[] | undefined {
  if (visited.has(name)) {
    return undefined;
  }
  if (visiting.has(name)) {
    const start = stack.indexOf(name);
    if (start < 0) {
      return undefined;
    }
    return [...stack.slice(start), name];
  }
  visiting.add(name);
  stack.push(name);
  for (const dependency of edges.get(name) ?? []) {
    if (!edges.has(dependency)) {
      continue;
    }
    const cycle = depthFirstSearch(dependency, edges, visiting, visited, stack);
    if (cycle) {
      return cycle;
    }
  }
  stack.pop();
  visiting.delete(name);
  visited.add(name);
  return undefined;
}

File: rubygems/pray-cli/lib/pray/dependency_graph.rb
# frozen_string_literal: true

module Pray
  module DependencyGraph
    module_function

    def find_dependency_cycle(edges)
      visiting = {}
      visited = {}
      stack = []
      edges.keys.sort.each do |name|
        next if visited[name]

        cycle = depth_first_search(name, edges, visiting, visited, stack)
        return cycle if cycle
      end
      nil
    end

    def depth_first_search(name, edges, visiting, visited, stack)
      return nil if visited[name]
      if visiting[name]
        start = stack.index(name)
        return nil unless start

        return stack[start..] + [name]
      end

      visiting[name] = true
      stack << name
      Array(edges[name]).each do |dependency|
        next unless edges.key?(dependency)

        cycle = depth_first_search(dependency, edges, visiting, visited, stack)
        return cycle if cycle
      end
      stack.pop
      visiting.delete(name)
      visited[name] = true
      nil
    end
  end
end
